irohad/torii/processor/impl/transaction_processor_impl.cpp

```diff
@@ -17,26 +17,46 @@
 
 #include "torii/processor/transaction_processor_impl.hpp"
 
+#include <boost/format.hpp>
+
 #include "backend/protobuf/transaction.hpp"
 #include "interfaces/iroha_internal/block.hpp"
 #include "interfaces/iroha_internal/proposal.hpp"
+#include "validation/stateful_validator_common.hpp"
 
 namespace iroha {
   namespace torii {
 
     using network::PeerCommunicationService;
 
+    static std::string composeErrorMessage(
+        const validation::TransactionError &tx_error) {
+      if (not tx_error.first.tx_passed_initial_validation) {
+        return (boost::format("Stateful validation error: transaction %s "
+                              "did not pass initial verification: "
+                              "checking '%s', error message '%s'")
+                % tx_error.second.hex() % tx_error.first.name
+                % tx_error.first.error)
+            .str();
+      }
+      return (boost::format("Stateful validation error in transaction %s: "
+                            "command '%s' with index '%d' did not pass "
+                            "verification with error '%s'")
+              % tx_error.second.hex() % tx_error.first.name
+              % tx_error.first.index % tx_error.first.error)
+          .str();
+    }
+
     TransactionProcessorImpl::TransactionProcessorImpl(
         std::shared_ptr<PeerCommunicationService> pcs,
         std::shared_ptr<MstProcessor> mst_processor)
         : pcs_(std::move(pcs)), mst_processor_(std::move(mst_processor)) {
       log_ = logger::log("TxProcessor");
 
-      // insert all txs from proposal to proposal set
+      // notify about stateless success
       pcs_->on_proposal().subscribe([this](auto model_proposal) {
         for (const auto &tx : model_proposal->transactions()) {
           auto hash = tx.hash();
```

Comment:
Use reference please to avoid copies, relevant snippet: https://godbolt.org/g/LBAj86